include/contraction/contractionGraph.hpp

```diff
@@ -40,220 +40,304 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
-     }
-
-     bool is_linear(V v) {
-         // Checking adjacent vertices constraint
-         auto adjacent_vertices = find_adjacent_vertices(v);
-
-         if (adjacent_vertices.size() == 2) {
-             // Checking u - v - w
-             V u = adjacent_vertices.front();
-             adjacent_vertices.pop_front();
-             V w = adjacent_vertices.front();
-             adjacent_vertices.pop_front();
-             if (is_shortcut_possible(u, v, w)) {
-                 return true;
-             }
-             return false;
-         }
-         return false;
-     }
+    typedef typename boost::graph_traits < G >::vertex_descriptor V;
+    typedef typename boost::graph_traits < G >::edge_descriptor E;
+    typedef typename boost::graph_traits < G >::edge_iterator E_i;
+    typedef typename boost::graph_traits < G >::out_edge_iterator EO_i;
+    typedef typename boost::graph_traits < G >::in_edge_iterator EI_i;
+
+    /*!
+        Prepares the _graph_ to be of type *directed*
+    */
+    explicit contractionGraph<G, t_directed>():
+        Pgr_base_graph<G, CH_vertex, CH_edge, t_directed>() {
+        min_edge_id = 0;
+    }
+
+    int64_t get_next_id() {
+        return --min_edge_id;
+    }
+
+    /*!
+        @brief get the vertex descriptors of adjacent vertices of *v*
+        @param [in] v vertex_descriptor
+        @return Identifiers<V>: The set of vertex descriptors adjacent
+        to the given vertex *v*
+    */
+    Identifiers<V> find_adjacent_vertices(V v) const {
+        EO_i out, out_end;
+        EI_i in, in_end;
+        Identifiers<V> adjacent_vertices;
+
+        for (
+            boost::tie(out, out_end) = out_edges(v, this->graph);
+            out != out_end;
+            ++out
+        )
+        adjacent_vertices += this->adjacent(v, *out);
+
+        for (
+            boost::tie(in, in_end) = in_edges(v, this->graph);
+            in != in_end;
+            ++in
+        )
+        adjacent_vertices += this->adjacent(v, *in);
+
+        return adjacent_vertices;
+    }
+
+    /*!
+        @brief vertices with at least one contracted vertex
+        @result The vids Identifiers with at least one contracted vertex
+    */
+    std::vector<E> get_shortcuts() {
+        pgrouting::Identifiers<E> eids;
+        E_i e, e_end;
+        for (boost::tie(e, e_end) = edges(this->graph); e != e_end; ++e) {
+            if (this->graph[*e].id < 0) {
+                eids += *e;
+                pgassert(!(this->graph[*e]).get_contracted_vertices().empty());
+            } else {
+                pgassert((this->graph[*e]).get_contracted_vertices().empty());
+            }
+        }
+        std::vector<E> o_eids(eids.begin(), eids.end());
+        std::sort(
+            o_eids.begin(),
+            o_eids.end(),
+            [&](E lhs, E rhs) {
+                return -1*(this->graph)[lhs].id < -1*(this->graph)[rhs].id;
+            });
```

Comment:
```suggestion
           std::sort(o_eids.begin(), o_eids.end(),
        [&graph = this->graph](const E& lhs, const E& rhs) {
            return graph[lhs].id > graph[rhs].id;
        });
```

- simplify the lambda using graph alias
- use const ref when possible
- is the -1 was useful?